README
Mozilla::DOM
------------

This module wraps the Mozilla DOM interface in Perl.
In conjuction with an embedded Gecko widget (e.g. Gtk2::MozEmbed),
you can use Perl to manipulate the browser DOM, handle DOM signals,
and create events such as mouse clicks -- all within a Mozilla-like
browser (so it also does JavaScript). See `perldoc Mozilla::DOM`
for more details. 

There are examples in the `examples' directory showing how
to use this module with Gtk2::MozEmbed. See the README file
in that directory. Check out Mozilla::Mechanize for a WWW::Mechanize-like
module that uses Mozilla::DOM. See `perldoc Mozilla::DOM::index` for
a list of all the POD files distributed with this module.


Dependencies
------------

You might have to manually install ExtUtils::Depends and ExtUtils::PkgConfig
before anything else. Sorry about that. You'll also have to install Mozilla
with its development headers (I think you can also just install the SDK).

Mozilla::DOM depends on:

  ExtUtils::Depends
  ExtUtils::PkgConfig
  `pkg-config --modversion mozilla-xpcom` >= 1.7
    or
  `pkg-config --modversion firefox-xpcom` >= 1.0

(I have no idea if Firefox works.)


Installation
------------

Note: I've only built this on Debian Linux with Mozilla 1.7.5
and 1.7.10. People running Gentoo with Mozilla 1.7.6 have reported
being able to build it.

To install Mozilla::DOM, run the following commands:

  perl Makefile.PL
  make           (this can take a while)
  make test      (not really any tests yet)
  make install   (as root)

(See "Dependencies" above.)

For this module to be useful, you'll probably also want
to build and install Gtk2::MozEmbed (> 0.03) *after* you
install this. That module includes Mozilla::DOM (through
the magic of ExtUtils::Depends) if it's already installed.
IMPORTANT: if you upgrade Mozilla::DOM, you also need
to rebuild and reinstall Gtk2::MozEmbed.


Experimental features
---------------------

Some features are added "experimentally". These are generally ones
that are either known not to be completely working, or I'm not sure
if they're really done right. I'll try to mark these clearly in the
documentation as experimental. To enable an experimental feature,
you pass an argument on the command line, for example

  perl Makefile.PL DEFINE="-DMDEXP_EVENT_LISTENER"

where "MDEXP" is just a (namespace-like) prefix and the rest
of the name identifies which feature is being enabled. Currently
the following features are experimental: EVENT_LISTENER.


Troubleshooting
---------------

A common problem seems to be that header files are put in different
directories for different installations of Mozilla. If so, you'll
get an error like

  mozilladom2perl.h:37:27: nsIWebBrowser.h: No such file or directory
  mozilladom2perl.h:38:26: nsISelection.h: No such file or directory

If this happens, do

  find . -name 'nsIWebBrowser.h'

from within the Mozilla include directory (e.g. /usr/include/mozilla).
It will probably be under a subdirectory of /usr/include/mozilla/.
Try adding that subdirectory to the set_inc method in Makefile.PL,
and please report it to me.

If you find that this module segfaults on Ubuntu, try this solution
from Boris Sukholitko which worked for me:
- add /usr/lib/firefox to /etc/ld.so.conf
- run /sbin/ldconfig
- remove /usr/lib/firefox/components/compreg.dat (regenerates itself)

On Jaunty, I've installed the libxul-unstable package, and it seems
they don't have either the directory where libxul.so exists in
/etc/ld.so.conf.d/ or symlinks there. As a hack, I did this....

  # cat >> libxulcrap.conf
  /usr/lib/xulrunner-devel-1.9.0.14/lib
  ctrl-d
  # ldconfig


Bugs
----

You can send bugs to me directly: Scott Lanning <slanning@cpan.org>


ChangeLog
0.22  2009-10-10 | updated to Gecko 1.9 (latest xulrunner/firefox)	 
  - patch (in 2008, sorry...) from Boris Sukholitko to make install
    work on Ubuntu Hardy
  - reworked install more
  - decided to only support Gecko 1.9 (I think), which means removing
    (see https://developer.mozilla.org/en/Gecko_1.9_Changes_affecting_websites )
	- NSEvent: preventBubble
	- NSRange: IntersectsNode, CompareNode, NSDetach
	- NSHTMLDocument: captureEvents, releaseEvents, routeEvent
  - if there are problems finding libxul.so, see the README

0.21  2007-06-07
  - Really sorry about the delay - in addition to me being lazy,
    gtkmozembed was segfaulting on Ubuntu!
    If you have this problem, see the Troubleshooting section of README.
  - patch from Boris Sukholitko for Mozilla::DOM::NSHTMLDocument open.
    Note: this only works with newer Firefox now, where apparently the API
    changed. I can add #ifdef to support the old version, but I'm not
    sure which version it changed in (let me know)
  - another patch from Boris for Mozilla::DOM::Window ScrollTo.
    (other Scroll* methods could be wrapped). While I was at it,
    I also wrapped GetScrollX, GetScrollY, ScrollBy, ScrollByLines,
    ScrollByPages (all but GetScrollbars, which is too hard)
  - minor edits of examples, docs; added Scroll example
  - really fixed META.yml no_index this time (I think),
    so hopefully modules in 'examples' directory aren't indexed on CPAN

0.20  2005-09-29
  - patch from José Auguste-Etienne for installing with Firefox on Mandriva
  - made certain methods perlish, by returning lists in list context:
    o  HTMLCollection
       - HTMLFormElement: GetElements
       - HTMLMapElement: GetAreas
       - HTMLTableElement: GetRows, GetTBodies
       - HTMLTableRowElement: GetCells
       - HTMLTableSectionElement: GetRows
       - NSHTMLDocument: GetEmbeds, GetPlugins
    o  HTMLOptionsCollection
       - HTMLSelectElement: GetOptions
    o  NamedNodeMap
       - Node: GetAttributes
       - DocumentType: GetEntities, GetNotations
    o  NodeList
       - Node: GetChildNodes
       - Document: GetElementsByTagName, GetElementsByTagNameNS
       - Element: GetElementsByTagName, GetElementsByTagNameNS
    o  WindowCollection
       - Window: GetFrames
    See the documentation of each method for details.

0.19  2005-09-27
  - added interfaces: Entity, Notation, NSRange, NSEvent, NSUIEvent,
    NSHTMLAnchorElement, NSHTMLAreaElement, NSHTMLButtonElement,
    NSHTMLDocument, NSHTMLFormElement, NSHTMLFrameElement, NSHTMLHRElement,
    NSHTMLImageElement, NSHTMLInputElement, NSHTMLOptionElement,
    NSHTMLSelectElement, NSHTMLTextAreaElement

0.18  2005-09-25
  - added interfaces:
    o  DocumentRange (for creating a Range, whose ToString allows
       something like get_text in HTML::TokeParser)
    o  NSHTMLElement (innerHTML (no outerHTML unfortunately))
    o  NSDocument (content-type, referer, etc.)
  - added Range (incomplete) and InnerHTML examples

0.17  2005-09-15
  - fixed constants: see DOMException (unsupported yet), Event, KeyEvent,
    MutationEvent, Node, Range, and WebNavigation (partially supported)
  - removed backwards-compatibility underscore_style aliases (see 0.06 below)

0.16  2005-09-05
  - added support for Firefox to Makefile.PL (copied from Gtk2::MozEmbed;
    untested, please report success or failure)
  - added no_index to META.yml to suppress indexing the examples on CPAN.
    Thanks to Adam Kennedy for pointing that out.

0.15  2005-09-04
  - moved POD from xs/DOM.xs into lib/Mozilla/DOM/*.pod. At the expense of
    the POD no longer being auto-generated, people can now read the POD
    without having to install the module (e.g. on CPAN), install time
    is slightly reduced, and I have better control over the POD output.
    (Still need to clean up the POD...)
    I also suppressed generating separate man pages because I don't see
    the point (you should delete any existing man/man3/Mozilla*.3pm files
    if you've installed this module before).
  - removed Glib dependency

0.14  2005-08-29
  - added most of WebNavigation, with an example WebNav.pm
  - require specifically enabling EventListener, using
    `perl Makefile.PL DEFINE="-DMDEXP_EVENT_LISTENER"`,
    after Torsten Schoenfeld reported that compiling Gtk2::MozEmbed
    was giving warnings (need help of a C++ expert)

0.13  2005-04-21
  - added browser object interfaces
    o  Navigator, History, Location, and Screen
    o  Window2 and WindowInternal, to get the browser objects
  - added the example BrowserObjects.pm
  - fixed the EventListener example

0.12  2005-04-20
  - keep in mind that Mozilla::DOM is alpha software, subject to change:
    o  added experimental EventListener support, with an example
       EventListeners.pm
    o deprecated all constants, although I didn't reimplement them
      correctly yet :/

0.11  2005-04-18
  - wrapped more interfaces (total is 88 now; there are >260 in Mozilla)
    o  54 HTML*Element classes
    o  HTMLCollection and HTMLOptionsCollection
  - added HTMLElements example
  - added the gendomxsubs.pl script which I use to generate XSUBs
    from Mozilla headers
  - fixed GetIID POD (but not yet the warnings during `make`)

0.10  2005-04-16
  - implemented Supports (QueryInterface method, plus GetIID class method
    for all interfaces), which means we can now create events!
  - added the Events example, showing mouse clicks and key presses
  - Uwe Voelker pointed me to the CPAN Forum. I registered to receive
    notification of posts, so you can post there if you want. See the
    'Contact' section of the README for details.

0.09  2005-04-11
  - added Elements example, demoing the tedium of DOM manipulation

0.08  2005-04-10
  - header includes - added two directories to set_inc in Makefile.PL:
    o  Torsten Schoenfeld noted that nsIWebBrowser.h is under the 'webbrwsr'
       directory on his system, so added that to the includes in Makefile.PL.

Makefile.PL
#!/usr/bin/perl
# xxx: I'd really like to make this more friendly to install....

use 5.008;
use strict;
use warnings;

use ExtUtils::MakeMaker;
use Config;
use Cwd qw(cwd);
use File::Spec;

my %pkg_to_try = (
    moz         => {
        pkgcfg => 'mozilla-xpcom',
        reqver => 1.7,
    },
    ff          => {
        pkgcfg => 'firefox-xpcom',
        reqver => 1.0,
    },
    mff         => {
        pkgcfg => 'mozilla-firefox-xpcom',
        reqver => 1.0,
    },
    xulunstable => {
        pkgcfg => 'libxul-unstable',
        reqver => 1.9,
    },
    xul         => {
        pkgcfg => 'libxul',
        reqver => 1.9,
    },
);
my @pkgs = qw/moz ff mff xulunstable xul/;
our %build_reqs = (
    'perl-ExtUtils-Depends'   => '0.205',
    'perl-ExtUtils-PkgConfig' => '1.07',
    (
        map { ( $pkg_to_try{$_}{pkgcfg} => $pkg_to_try{$_}{reqver} ) } @pkgs
    ),
);

unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
           . "1") {
    warn "$@\n";
    WriteMakefile(
        PREREQ_FATAL => 1,
        PREREQ_PM    => {
            'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
            'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
        },
    );
    exit 1; # not reached
}

my %pkgcfg = ExtUtils::PkgConfig->find(
    map { $_ . ' >= ' . $build_reqs{$_} } map { $pkg_to_try{$_}{pkgcfg} } @pkgs
);

my ($pkg) = $pkgcfg{pkg} =~ /^(\S+) /;

mkdir 'build', 0777;

ExtUtils::PkgConfig->write_version_macros (
    "build/mozilladom2perl-version.h",
    $pkg => 'MOZ_DOM',
);

my $incdir = `pkg-config --variable=includedir $pkg`;
my $libdir = `pkg-config --variable=libdir $pkg`;
chomp($incdir, $libdir);

unless ($libdir) {
    $libdir = `pkg-config --variable=sdkdir $pkg`;
    chomp($libdir);
}

my $mozdom = ExtUtils::Depends->new('Mozilla::DOM');

# -I/usr/include/mozilla -I/usr/include/mozilla/xpcom
# -I/usr/include/mozilla/string -I/usr/include/mozilla/nspr
$mozdom->set_inc($pkgcfg{cflags}, '-I.', '-I./build',
                 "-I$incdir", "-I$incdir/dom", "-I$incdir/webbrwsr",
                 "-I$incdir/content", "-I$incdir/docshell");

# -L/usr/lib/mozilla -lxpcom -lplds4 -lplc4 -lnspr4 -ldl -lc -lpthread
$mozdom->set_libs($pkgcfg{libs});

$mozdom->add_xs(<xs/*.xs>);
# XXX: should move DOM.pm into lib/Mozilla/ then use PMLIBDIRS in WriteMakefile instead
$mozdom->add_pm(
    'DOM.pm' => '$(INST_LIBDIR)/DOM.pm',
    map {   # .pod and .pm files in lib/Mozilla/DOM/
        my $inst = $_;
        $inst =~ s{lib/Mozilla}{};
        $_ => "\$(INST_LIBDIR)$inst";
    } glob('lib/Mozilla/DOM/*.{pod,pm}'),
);
$mozdom->add_typemaps(map {File::Spec->catfile(cwd(), $_)} ('mozilladom.typemap'));

$mozdom->install('mozilladom2perl.h',
                 # 'build/mozilladom2perl-autogen.h',
                 'build/mozilladom2perl-version.h',
                 'doctypes');
$mozdom->save_config('build/IFiles.pm');

# xxx:
# https://developer.mozilla.org/en/Gecko_1.9_Changes_affecting_websites#intersectsNode_has_been_removed
my @defines = ();
push @defines, '-DMD_XUL'
  if $pkg eq $pkg_to_try{xul}{pkgcfg} or $pkg eq $pkg_to_try{xulunstable}{pkgcfg};

WriteMakefile(
    NAME          => 'Mozilla::DOM',
    VERSION_FROM  => 'DOM.pm',
    ABSTRACT_FROM => 'DOM.pm',
    XSPROTOARG    => '-noprototypes',
    MAN3PODS      => {},    # don't create man pages
    LD            => "LD_RUN_PATH=$libdir $Config{ld}",
    CC            => 'g++',
    XSOPT         => '-C++',
    dist => {
        # don't index examples directory on CPAN
        PREOP => 'echo -e "no_index:\n  directory:\n    - examples" >> ${DISTVNAME}/META.yml',
    },
    $mozdom->get_makefile_vars,

    (@defines ? ('DEFINE' => join(' ', @defines)) : ()),
);


package MY;

# Copied from Glib::MakeHelper
# (see NOTICE in `perldoc Glib::MakeHelper` on const_cccmd)
sub const_cccmd {
        my $inherited = shift->SUPER::const_cccmd(@_);
        return '' unless $inherited;
        use Config;
        $inherited .= ($Config{cc} eq 'cl') ? ' /Fo$@' : ' -o $@';
        return $inherited;
}

# Copied from Glib::MakeMaker `postamble_clean'
sub postamble {
'
realclean ::
	-$(RM_RF) build blib_done perl-$(DISTNAME).spec
'
}
